src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseConcreteTypeAnalyzer.cs

```diff
@@ -0,0 +1,232 @@
+            {
+                if (coll.ParameterAssignments.TryGetValue(parameter, out var assignments))
+                {
+                    if (parameter.ContainingSymbol is IMethodSymbol method)
+                    {
+                        if (CanUpgrade(method))
+                        {
+                            Report(parameter, parameter.Type, assignments, UseConcreteTypeForParameter);
+                        }
+                    }
+                }
+            }
+
+            foreach (var pair in coll.MethodReturns)
+            {
+                var method = pair.Key;
+                var returns = pair.Value;
+
+                if (CanUpgrade(method))
+                {
+                    Report(method, method.ReturnType, returns, UseConcreteTypeForMethodReturn);
+                }
+            }
+
+            void Report(ISymbol sym, ITypeSymbol fromType, ConcurrentBag<ITypeSymbol> assignments, DiagnosticDescriptor desc)
+            {
+                using var types = PooledHashSet<ITypeSymbol>.GetInstance(SymbolEqualityComparer.Default);
+                types.UnionWith(assignments);
+
+                var assignedNull = types.Remove(coll.Void!);
+
+                if (types.Count == 1)
+                {
+                    var toType = types.Single();
+                    if (assignedNull)
+                    {
+                        toType = toType.WithNullableAnnotation(Analyzer.Utilities.Lightup.NullableAnnotation.Annotated);
+                    }
+
+                    if (fromType.TypeKind == TypeKind.Interface)
+                    {
+                        if (!toType.AllInterfaces.Contains(fromType, SymbolEqualityComparer.Default))
+                        {
+                            // If the from type is an interface, the destination type must implement that interface
+                            // This deals with the case of going from T to Task<T> for example.
+                            return;
+                        }
+                    }
+                    else
+                    {
+                        if (!toType.DerivesFrom(fromType))
+                        {
+                            // if the from type is a class, then the to type must be more derived than the
+                            // class
+                            return;
+                        }
+                    }
+
+                    if (toType.TypeKind == TypeKind.Class
+                        && !SymbolEqualityComparer.Default.Equals(fromType, toType)
+                        && toType.SpecialType != SpecialType.System_Object
+                        && toType.SpecialType != SpecialType.System_Delegate)
+                    {
+                        var fromTypeName = GetTypeName(fromType);
+                        var toTypeName = GetTypeName(toType);
+                        var diagnostic = Diagnostic.Create(desc, sym.Locations[0], sym.Name, fromTypeName, toTypeName);
```

Comment:
Consider using the following extension:

https://github.com/dotnet/roslyn-analyzers/blob/2b5a3b2f8e5e841c18cde4855a22cf8161ad4986/src/Utilities/Compiler/Extensions/DiagnosticExtensions.cs#L76-L82

```suggestion
                        var diagnostic = sym.CreateDiagnostic(desc, sym.Name, fromTypeName, toTypeName);
```

Reply:
Done.